src/gen_rules.ml

```diff
@@ -1591,40 +1618,72 @@ module Gen(P : Params) = struct
-             corresponding .ml file.\n\
-             Modules without an implementation are not recommended, \
-             see this discussion:\n\
-             \n\
-            \  https://github.com/janestreet/jbuilder/issues/9\n\
-             \n\
-             In the meantime I'm setting up a rule for copying %s to %s.\n"
-            name (Path.to_string dir)
-            mli_fname ml_fname;
-          let dir = Path.append ctx.build_dir dir in
-          add_rule
-            (Build.copy
-               ~src:(Path.relative dir mli_fname)
-               ~dst:(Path.relative dir ml_fname));
-          ml_fname
-        | Some ml_fname -> ml_fname
-      in
-      Some
-        { Module.
-          name
-        ; ml_fname  = ml_fname
-        ; mli_fname = mli_fname
-        ; obj_name  = ""
-        })
+    let ml_impls = parse_one_set ml_files  in
+    let ml_intfs = parse_one_set mli_files in
+    let re_impls = parse_one_set re_files  in
+    let re_intfs = parse_one_set rei_files in
+    let to_module impls intfs =
+      String_map.merge impls intfs ~f:(fun name impl_fname intf_fname ->
+        let impl_fname =
+          match impl_fname with
+          | None ->
+            let intf_fname = Option.value_exn intf_fname in
+            let impl_fname = String.sub intf_fname ~pos:0 ~len:(String.length intf_fname - 1) in
+            Format.eprintf
+              "@{<warning>Warning@}: Module %s in %s doesn't have a \
+               corresponding .ml file.\n\
+               Modules without an implementation are not recommended, \
+               see this discussion:\n\
+               \n\
+              \  https://github.com/janestreet/jbuilder/issues/9\n\
+               \n\
+               In the meantime I'm setting up a rule for copying %s to %s.\n"
+              name (Path.to_string dir)
+              intf_fname impl_fname;
+            let dir = Path.append ctx.build_dir dir in
+            add_rule
+              (Build.copy
+                 ~src:(Path.relative dir intf_fname)
+                 ~dst:(Path.relative dir impl_fname));
+            impl_fname
+          | Some impl_fname -> impl_fname
+        in
+        Some (Module.create ~name ~impl_fname ?intf_fname ())
+      ) in
+    let ml_modules = to_module ml_impls ml_intfs in
+    let reason_modules = to_module re_impls re_intfs in
+    let ml_re_collisions =
+      String_map.merge ml_modules reason_modules ~f:(fun _ m r ->
+        match m, r with
+        | Some m, Some r -> Some (m, r)
+        | Some _, None
+        | None, Some _ -> None
+        | None, None -> assert false
+      ) in
+    if String_map.is_empty ml_re_collisions then
+      String_map.merge ml_modules reason_modules ~f:(fun _ m r ->
+        match m, r with
+        | Some m, None
+        | None, Some m -> Some m
+        | None, None
+        | Some _, Some _ -> assert false
+      )
+    else
+      die "The following modules have both reason and ocaml sources:@.%a"
```

Comment:
Indeed, we could just look at the extension, but I think it would be cleaner to keep separated flags in [impl_fname] and [intf_fname]. We could change the types in module.mli as follow:

```ocaml
module Syntax : sig
  type t = OCaml | Reason
end

module File : sig
  type t =
    { name   : string
    ; syntax : Syntax.t
    }
end

type t =
  { name : string
  ; impl : File.t
  ; intf : File.t option
  ...
```